torch/onnx/verification.py

```diff
@@ -1070,25 +1105,82 @@ def pretty_print(self):
         total_rows = self._total_rows()
         for line in range(total_rows):
             print(self._str_at_line(line).rstrip())
-        # Summarize leaf subgraphs with mismatch.
-        print(" Mismatch leaf subgraphs: ".center(80, "="))
-        print(
-            [
-                graph_info.id
-                for graph_info in self.graph_info.all_mismatch_leaf_graph_info()
-            ]
+        if self.graph_info.has_mismatch():
+            # Summarize leaf subgraphs with mismatch.
+            print(" Mismatch leaf subgraphs: ".center(80, "="))
+            print(
+                [
+                    graph_info.id
+                    for graph_info in self.graph_info.all_mismatch_leaf_graph_info()
+                ]
+            )
+            # Summarize node kinds with mismatch.
+            mismatch_node_kinds: Dict[str, int] = {}
+            for graph_info in self.graph_info.all_mismatch_leaf_graph_info():
+                node_kinds = graph_info.essential_node_kinds()
+                if len(node_kinds) == 1:
+                    node_kind = node_kinds.pop()
+                    mismatch_node_kinds[node_kind] = (
+                        mismatch_node_kinds.get(node_kind, 0) + 1
+                    )
+            print(" Mismatch node kinds: ".center(80, "="))
+            print(mismatch_node_kinds)
+        else:
+            print(" No mismatch found. ".center(80, "="))
+
+
+class OnnxTestCaseRepro:
+    def __init__(self, repro_dir):
+        self.repro_dir = repro_dir
+        self.proto, self.inputs, self.outputs = onnx_proto_utils.load_test_case(
+            repro_dir
         )
-        # Summarize node kinds with mismatch.
-        mismatch_node_kinds: Dict[str, int] = {}
-        for graph_info in self.graph_info.all_mismatch_leaf_graph_info():
-            node_kinds = graph_info.essential_node_kinds()
-            if len(node_kinds) == 1:
-                node_kind = node_kinds.pop()
-                mismatch_node_kinds[node_kind] = (
-                    mismatch_node_kinds.get(node_kind, 0) + 1
-                )
-        print(" Mismatch node kinds: ".center(80, "="))
-        print(mismatch_node_kinds)
+
+    @_beartype.beartype
+    @classmethod
+    def create_test_case_repro(
+        cls, proto: bytes, inputs, outputs, dir: str, name: Optional[str] = None
+    ):
+        """Create a repro under "{dir}/test_{name}" for an ONNX test case.
+
+        Args:
+            proto: ONNX model proto.
+            inputs: Inputs to the model.
+            outputs: Outputs of the model.
+            dir: Directory to save the repro.
+            name: Name of the test case. If not specified, a name based on current time
+                will be generated.
+        Returns:
+            Path to the repro.
+        """
+        if name is None:
+            name = datetime.datetime.now().strftime("%Y_%m_%d_%H_%M_%S_%f")
```

Comment:
Add a random string after the datetime maybe?

Reply:
😄 This is getting to microseconds .. Should be safe enough?